File: guests/ocaml/scroll.ml
(* The scroll conformance scene, OCaml port — the viewport grows so
   the enclosing track constrains it (an unconstrained viewport hugs
   its content and nothing overflows); the bottom button, reachable
   only by scrolling, proves the scrolled-to content is live. See
   guests/rust/scroll.rs and tools/scenes/scroll.steps. *)

open Kaya_wire
open Kaya_app

let () =
  let app = Kaya_app.create () in

  let status = ref None in
  build app
    (let* () = window_title "scroll" in
     let* s = signal (Str "at top") in
     status := Some s;
     let on_bottom tx = write s (Str "bottom clicked") tx in
     let row i =
       let* caption = signal (Str (Printf.sprintf "row %d" i)) in
       label ~bind:caption ()
     in
     let* root =
       column
         [
           label ~bind:s () (* label#0 *);
           scroll ~grow:1.0
             (column
                (List.init 29 (fun i -> row (i + 1))
                @ [ button ~text:"bottom" ~on_click:on_bottom () (* button#0 *) ]));
         ]
     in
     mount root);

  ignore !status;

  exit (run app)

File: guests/swift/scroll.swift
// The scroll scene, Swift port — guests/rust/scroll.rs,
// tools/scenes/scroll.steps.

import Foundation

let app = KayaApp()

app.build { tx in
    tx.window(title: "scroll")
    let status = tx.signal(.str("at top"))
    let root = tx.column { root in
        tx.label(bind: status)  // label#0
        tx.scroll(grow: 1) { _ in  // scroll#0
            tx.column { _ in
                for i in 1...29 {
                    let caption = tx.signal(.str("row \(i)"))
                    tx.label(bind: caption)
                }
                tx.button(
                    "bottom",
                    onClick: { inner in  // button#0
                        inner.write(status, .str("bottom clicked"))
                    })
            }
        }
        return root
    }
    tx.mount(root)
}

app.run()
